File: State/python/state.py
from abc import ABC, abstractmethod


class StateAbstract(ABC):
    @abstractmethod
    def buy(self, cost: float):
        pass

    @abstractmethod
    def deposit(self, amount: float):
        pass

    @abstractmethod
    def setContext(self, context):
        pass


class Context:

    def __init__(self, initState: StateAbstract):
        self.__residue: float = 0.0
        self.setState(initState)

    @property
    def residue(self):
        return self.__residue

    def setState(self, state: StateAbstract):
        self.__state: StateAbstract = state
        print(self.__state)
        self.__state.setContext(self)

    def buySomething(self, cost: float):
        self.__state.buy(cost)

    def depositSalary(self, amount: float):
        self.__state.deposit(amount)

    def _reduce(self, cost: float):
        self.__residue -= cost

    def _add(self, amount: float):
        self.__residue += amount


class IsDeudor(StateAbstract):

    def __init__(self):
        self.__context = None

    def setContext(self, context: Context):
        self.__context = context

    def buy(self, cost: float):
        print("No puedes comprar eres un deudor")

    def deposit(self, amount: float):
        self.__context._add(amount)
        self.__context.setState(NoDeudor())
        print(f'tengo de saldo total {self.__context.residue}')


class NoDeudor(StateAbstract):
    def __init__(self):
        self.__context = None

    def setContext(self, context: Context):
        self.__context = context

    def buy(self, cost: float):
        if(self.__context.residue >= cost):
            self.__context._reduce(cost)
            print(
                f'Realizo una compra de {cost} me queda un saldo total de {self.__context.residue}')
        else:
            print(
                f'Quieres gastar {cost} con tu saldo total de {self.__context.residue}')

        if(self.__context.residue == 0):
            self.__context.setState(IsDeudor())

    def deposit(self, amount: float):
        self.__context._add(amount)
        print(f'tengo de saldo total {self.__context.residue}')


class NewCount(StateAbstract):
    def __init__(self):
        self.__context = None

    def setContext(self, context: Context):
        self.__context = context

    def buy(self, cost: float):
        print('Su cuenta es nueva necesita hacer un deposito inicial primero')

    def deposit(self, amount: float):
        self.__context._add(amount)
        print(f'tengo de saldo total {self.__context.residue}')
        self.__context.setState(NoDeudor())


if __name__ == '__main__':
    context = Context(NewCount())
    context.buySomething(50)
    context.depositSalary(100)
    context.buySomething(50)
    context.buySomething(50)
    context.buySomething(50)
    context.depositSalary(50)
    context.buySomething(55)
    context.buySomething(50)

File: State/typeScript/state.ts
class Bank {
  private _instanceState!: IStateAccount;
  private _amount: number = 0;

  constructor(initialState: IStateAccount) {
    this.setState(initialState);
  }

  public get amount(): number {
    return this._amount;
  }
  public set amount(amount: number) {
    this.amount = amount;
  }

  public setState(state: IStateAccount): void {
    console.log(`ahora el estado es  ${(<any>state).constructor.name}`);
    this._instanceState = state;
    this._instanceState.setContext(this);
  }

  buy(amount: number) {
    this._instanceState.buy(amount);
  }

  addAmount(amount: number) {
    this._instanceState.addAmount(amount);
  }

  discount(amount: number) {
    this._amount -= amount;
  }

  deposit(amount: number) {
    this._amount += amount;
  }
}

interface IStateAccount {
  buy(mount: number): void;
  addAmount(mount: number): void;
  setContext(context: Bank): void;
}

class newAccount implements IStateAccount {
  private _context!: Bank;

  setContext(context: Bank) {
    this._context = context;
  }

  addAmount(mount: number): void {
    this._context.deposit(mount);
    this._context.setState(new NoDeudor());
  }

  buy(mount: number): void {
    console.log('No tienes dinero, primero has un deposito');
  }
}

class NoDeudor implements IStateAccount {
  private _context!: Bank;

  buy(mount: number): void {
    if (this._context.amount >= mount) {
      console.log(
        `tengo en el banco ${this._context.amount} y voy a gastar ${mount}`
      );
      this._context.discount(mount);
    } else {
      console.log(
        `no puedes gastar mas de lo que tienes en el banco ${this._context.amount} y quieres gastar ${mount}`
      );
    }

    if (this._context.amount == 0) {
      this._context.setState(new IsDeudor());
    }
  }

  addAmount(mount: number): void {
    this._context.deposit(mount);
    console.log(`ya tienes saldo en tu cuenta :) = ${this._context.amount}`);
  }
  setContext(context: Bank): void {
    this._context = context;
  }
}

class IsDeudor implements IStateAccount {
  private _context!: Bank;

  buy(mount: number): void {
    console.log('no tienes saldo suficiente');
  }
  addAmount(mount: number): void {
    this._context.deposit(mount);
    console.log(`ya tienes saldo en tu cuenta :) = ${this._context.amount}`);
    this._context.setState(new NoDeudor());
  }
  setContext(context: Bank): void {
    this._context = context;
  }
}

const bank = new Bank(new newAccount());
bank.buy(19);
bank.addAmount(100);
bank.buy(10);
bank.buy(100);
bank.addAmount(200);
bank.buy(100);
bank.addAmount(100);
bank.buy(290);
bank.buy(12);
bank.addAmount(100);
bank.buy(100)
